### ISSUE TITLE
Building a package or workspace in a subdirectory

### ISSUE THREAD

**Issue by paulyoung** [ipetkov/crane#209]
I have a directory structure that looks like this:

```
.
├── backend
│   ├── Cargo.lock
│   ├── Cargo.toml
...
├── frontend
...
```

Up until now I've been able to set `src = ./backend` but now I need to access some files that live in the root.

I've gotten as far as this:

```nix
{
  src = ./.;
  cargoToml = ./backend/Cargo.toml;
  cargoLock = ./backend/Cargo.lock;
  cargoExtraArgs = "--manifest-path ./backend/Cargo.toml";
}
```

but now I seem to be stuck because `cargo manifest` doesn't get `cargoExtraArgs`: https://github.com/ipetkov/crane/blob/0017a8da6824c5f27c48cce7d614d3e590e75e35/lib/setupHooks/installFromCargoBuildLogHook.sh#L34

Is there a better way to approach this?

**Comment by dpc** [comment]
Best way IMO, is the filter `src` with an explicit list of stuff to include. Example from our project: https://github.com/fedimint/fedimint/blob/65b2058e52b773f83e023a7cbb1d13988bc19197/flake.nix#L424

**Comment by paulyoung** [comment]
If I understand correctly, I think you’re suggesting using a filter to make it look like the contents of the `backend` directory (in my case) are at the root. Is that correct?

If not, could you help me understand how filtering would help? The issue seems to be that `Cargo.toml` isn’t at the root.

**Comment by paulyoung** [comment]
> If I understand correctly, I think you’re suggesting using a filter to make it look like the contents of the `backend` directory (in my case) are also at the root. Is that correct?

I think this approach would be problematic anyway because it would require changing relative paths that are looking for other files at the root and then that wouldn’t work in a dev shell when using `cargo` directly.

**Comment by paulyoung** [comment]
I think I’m going to see if introducing `cargoManifestExtraArgs` (which would default to `cargoExtraArgs`) fixes my issue and then maybe make a PR.

**Comment by dpc** [comment]
> If I understand correctly, I think you’re suggesting using a filter to make it look like the contents of the `backend` directory (in my case) are also at the root. Is that correct?

No, I was going to suggest making them a workspace, and just filtering necessary files.

**Comment by paulyoung** [comment]
I already have a workspace in `./backend` and build specific crates using `-p` so not sure how that helps. I'm trying to say `src = ./.` and still have `crane` be able to find what it needs without changing any of the existing directory structure.

I've been using `naersk` for a while and it differentiates between `src` and `root` for this purpose, although it doesn't work as it's supposed to at the moment.

**Comment by ipetkov** [comment]
> I'm trying to say src = ./. and still have crane be able to find what it needs without changing any of the existing directory structure.

@paulyoung Have you tried setting `sourceRoot = "source/backend";` and building that way? It's probably a much simpler solution to change into the directory which contains the workspace root (and still be able to access the parent directory).

If your build needs to work from the parent directory, you can always update the `installPhaseCommand` to temporarily move into the directory:

```nix
buildPackage {
  # other settings...
  installPhaseCommand = ''
    pushd ./backend
    installFromCargoBuildLog "$out" "$cargoBuildLog"
    popd
  '';
}
```

**Comment by paulyoung** [comment]
@ipetkov I haven’t (I wasn’t aware I could do that) but that sounds like a better approach. I’ll give it a try.

I was hoping there’d be a better way than what I’ve tried so far.

**Comment by ipetkov** [comment]
@paulyoung I believe this is probably the simplest way to achieve it. Also made things a bit easier and documented the approach in https://github.com/ipetkov/crane/pull/212 in case you want to give that a try!

**Comment by paulyoung** [comment]
Using `sourceRoot` instead gets me to the same place as what I was doing with #210 but with either approach the parent directory of the workspace isn't what I'd expect it to be.

I have a test which ascends to a parent directory with `../../..`. When I run this test with `cargo` it finds my root directory containing my `backend` directory and some other files that I want to use during the test.

When I run this with Nix, the directory I end up in contains:

```
.
|- tmp.xYwl96bE4K
|- .sandbox.sb
|- source
'- env-vars
```

I can provide an example if necessary.

**Comment by paulyoung** [comment]
Using `installPhaseCommand` instead of `sourceRoot` results in the same thing.

**Comment by ipetkov** [comment]
> I can provide an example if necessary.

An example would be great!

It does look like your build script is hitting the root of the builder sandbox itself (the `source` directory should contain the actual unpacked version of the `src` input to the derivation). I'm not sure if maybe the way Nix cleans the sources it ends up omitting one parent directory or something. I think being able to look at your specific configs would help!

**Comment by paulyoung** [comment]
@ipetkov please see https://github.com/paulyoung/crane-nested-issue

**Comment by paulyoung** [comment]
Actually, I did something different in that repo (I found that I had to use `craneLib.cleanCargoSource` this time or I got an error) and that appears to be able to reach the root.

The file I'm interested in is missing but that's probably because of `cleanCargoSource`

**Comment by paulyoung** [comment]
On that, the error I was getting was `chmod: cannot access 'source/nested': No such file or directory`

I wasn't getting that in my actual repo but may have been on a slightly older version.

**Comment by ipetkov** [comment]
Thanks for the minimal example @paulyoung, I've figured out the issue!

So turns out that the generic unpack phase will create a directory based on the _name of the derivation_ that is passed in:
- When using `src = pkgs.lib.cleanSource ./.;` (which `cleanCargoSources` also does) the source name defaults to `source` (though it is possible to override this).
- When using `src = ./nested;` the source name defaults to "nested"
- When using `src = ./.;` the source name defaults to `<store-hash>-source`

So in pretty much all the cases except the first trying to set `sourceRoot = "source/whatever"` fails since the root of the source isn't named `source`!

There's pretty much a different workaround we need to use (and I'll update the docs with this) which uses a postPatch hook to cd in there without having to guess what the source is called. Then things should work whether or not a source filter is used!

```nix
# Changing the package definition to this works for me:
craneLib.buildPackage {
  inherit cargoArtifacts;
  pname = "foo";
  src = ./.;
  postUnpack = ''
    cd $sourceRoot/nested
    sourceRoot="."
  '';
  cargoToml = ./nested/Cargo.toml;
  cargoLock = ./nested/Cargo.lock;
  nativeBuildInputs = [
    pkgs.libiconv
  ];
  doCheck = true;
}
```

**Comment by paulyoung** [comment]
Thanks. That definitely fixes the standalone example but still isn't working for me in my real project. I'll keep trying to reproduce the issue.

**Comment by paulyoung** [comment]
<blockquote>
When I run this with Nix, the directory I end up in contains:
<br/><br/>
<pre>
.
|- .sandbox.sb
|- env-vars
|- source
'- tmp.xYwl96bE4K
</pre>
</blockquote>

I'm seeing the above when using `src = ./.;` and that doesn't seem to line up with what you said about how things should be named so wondering if that's somehow related to the issue I'm having.

**Comment by ipetkov** [comment]
Not sure, what version of Nix are you using?

Also in case you weren't familiar with this flag, you can run `nix build .# --keep-failed` to keep the directory after the builder exits so you can explore it interactively

**Comment by paulyoung** [comment]
Thanks, I’ll try that. I was looking for the equivalent of `-K`.

**Comment by paulyoung** [comment]
I think this just confirms what I already knew.

```
note: keeping build directory '/private/tmp/nix-build-rust-workspace-unknown.drv-1'
```

```
$ ls /private/tmp/nix-build-rust-workspace-unknown.drv-1
env-vars  source  tmp.W3JFsPQ3WN
```

```
$ ls /private/tmp/nix-build-rust-workspace-unknown.drv-1/source
Cargo.lock  Cargo.toml	src  target
```
